src/maybe.ts

```diff
@@ -792,4 +792,42 @@ export const toString = <T>(maybe: Maybe<T>): string => {
   return `${maybe.variant}${body}`;
 };
 
+/** A lightweight object defining how to handle each variant of a Maybe. */
+type Matcher<T, A> = {
+  Just: (value: T) => A;
+  Nothing: () => A;
+};
+
+/**
+  Performs the same basic functionality as `getOrElse`, but instead of simply
+  unwrapping the value if it is `Just` and applying a value to generate the same
+  default type if it is `Nothing`, lets you supply functions which may transform
+  the wrapped type if it is `Just` or get a default value for `Nothing`.
+
+  This is kind of like a poor-man's version of pattern matching, which
+  JavaScript currently lacks.
```

Comment:
In the spirit of making the potential use cases clear, it might be nice to have an example of what this _replaces_.

```js
// instead of
let result;
if (isJust(foo)) {
  result = foo;
} else {
  result = 'a default value'
}

// you can do
const result = match({
  Just: (x) => x,
  Nothing: 'a default value'
}, foo);
```

Reply:
Making it clearer is a big win; though I'd never write it like that first example. I'd write that as a ternary every time.

```ts
const result = isJust(foo) ? unwrap(foo) : 'a default value';
```

Reply:
Sure, _you_ wouldn't, but that's pretty common code in the wild.